### PR TITLE
[MRG] speed similar rules computation

### PR DESCRIPTION
**Pull request by datajms**

### CODE REVIEW

File: examples/plot_credit_default.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+
+
+
+
+###############################################################################
+# Data import and preparation
+# ..................
+#
+# There are 3 categorical variables (SEX, EDUCATION and MARRIAGE) and 20 numerical variables.
+# The target (credit defaults) is transformed in a binary variable with integers 0 (no default) and 1 (default).
```
Comment (ngoix) [inline examples/plot_credit_default.py]: rm "with integers"?

File: examples/plot_credit_default.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    archive = RemoteFileMetadata(
+        filename='default of credit card clients.xls',
+        url='https://archive.ics.uci.edu/ml/machine-learning-databases/00350/default%20of%20credit%20card%20clients.xls',
+        checksum=('30c6be3abd8dcfd3e6096c828bad8c2f'
+                  '011238620f5369220bd60cfc82700933'))
+
+    if not exists(join(sk_data_dir, archive.filename)):
+        _fetch_remote(archive, dirname=sk_data_dir)
+    
+    return(pd.read_excel(join(sk_data_dir, archive.filename), sheetname='Data', header=1))
```
Comment (ngoix) [inline examples/plot_credit_default.py]: do you have to return a pandas dataframe here? would be better to return a numpy array instead...

Comment (ngoix) [reply examples/plot_credit_default.py]: or a `Bunch` from sklearn/utils

File: examples/plot_credit_default.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+# Importing data
+data = _load_credit_data()
+
+# Shuffling data, preparing target and variables
+data = data.sample(frac = 1).reset_index(drop=True)
```
[the remark this answers is not in the document]
Comment (ngoix) [reply examples/plot_credit_default.py]: you can use the shuffle method from sklearn/utils

File: examples/plot_credit_default.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+                  param_grid={
+                  'max_depth': range(2, 7, 1),
+                  'max_features': np.linspace(0.1, 0.5, 5)
+                  },
+                  scoring = {'AUC': 'roc_auc'}, cv = 5, refit = 'AUC', n_jobs = -1)
+#RF = RandomForestClassifier()
+RF.fit(X_train, y_train)
+scoring_RF = RF.predict_proba(X_test)[:, 1]
+
+print("Decision Tree parameters : "+str(DT.best_params_))
```
Comment (ngoix) [inline examples/plot_credit_default.py]: "Decision Tree selected parameters"

File: examples/plot_credit_default.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+print("Random Forest parameters : "+str(RF.best_params_))
```
Comment (ngoix) [inline examples/plot_credit_default.py]: idem

File: examples/plot_credit_default.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+###############################################################################
+# Above are printed the parameters which have been chosen by the Grid Search.
```
Comment (ngoix) [inline examples/plot_credit_default.py]: then you can remove these two lines

File: examples/plot_credit_default.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+curves = [roc_curve, precision_recall_curve]
+xlabels = ['False Positive Rate', 'Recall (True Positive Rate)']
+ylabels = ['True Positive Rate (Recall)', 'Precision']
+
+for ax, curve, xlabel, ylabel in zip(axes.flatten(), curves,
+                                            xlabels, ylabels):
+    if curve == precision_recall_curve:
+        y_rf1, x_rf1, _ = curve(y_test, scoring_DT)
+        y_rf2, x_rf2, _ = curve(y_test, scoring_RF)
```
Comment (ngoix) [inline examples/plot_credit_default.py]: x_dt  x_rf instead of x_rf1 x_rf2?

File: examples/plot_credit_default.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+###############################################################################
+# Above are printed the parameters which have been chosen by the Grid Search.
+
+
+
+# Plot ROC and PR curves
+
+fig, axes = plt.subplots(1, 2, figsize=(12, 5),
+                         sharex=True, sharey=True)
+matplotlib.rcParams.update({'font.size': 18})
```
Comment (ngoix) [inline examples/plot_credit_default.py]: why do you need this?

File: examples/plot_credit_default.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+                   feature_names=feature_names, recall_min = 0.05, precision_min = 0.6
+                  )
+clf.fit(X_train, y_train)
+scoring = clf.decision_function(X_test)
+
+print(str(len(clf.rules_)) + ' rules have been built.')
+print(clf.rules_[:5])
+
+###############################################################################
+# The most precise rules are displayed above.
```
Comment (ngoix) [inline examples/plot_credit_default.py]: put it inside the print, not in comments.

File: examples/plot_credit_default.py [hunk 10]
```diff
@@ -0,0 +1,318 @@
+"""
+============================================
```
Comment (ngoix) [inline examples/plot_credit_default.py]: one more "=" here and below

File: examples/plot_credit_default.py [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        label = ('FraudToRules, AUC: %0.3f' % auc(x_rf1, y_rf1))
+        ax.scatter(x_rf1, y_rf1, c='b', s=10, label=label)
+        label = ('Random Forest, AUC: %0.3f' % auc(x_rf2, y_rf2))
+        ax.plot(x_rf2, y_rf2, '-.', lw=1, label=label, c = 'g')
+        label = ('ExtraTrees, AUC: %0.3f' % auc(x_rf3, y_rf3))
+        ax.scatter(x_rf3, y_rf3, c='r', s=10, label=label)
+        ax.set_title("ROC Curves", fontsize=20)
+        ax.legend(loc='upper center', fontsize = 8) 
+
+    ax.set_xlabel(xlabel)
```
Comment (ngoix) [inline examples/plot_credit_default.py]: add `fontsize=18` here instead of using rcParams

File: examples/plot_credit_default.py (hunk 8 above)
Comment (ngoix) [reply examples/plot_credit_default.py]: (you can put `fontsize=18` directly inside your `set_xlabel` / `set_ylabel` below)

File: examples/plot_credit_default.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.model_selection import GridSearchCV
+from sklearn.tree import DecisionTreeClassifier
+from ftrules import FraudToRules
+
+rng = np.random.RandomState(42)
+
+
+def _load_credit_data():
+    import sys
```
Comment (ngoix) [inline examples/plot_credit_default.py]: do you need this import?

File: examples/plot_credit_default.py [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+X_train = data[:n_samples_train]
+X_test = data[n_samples_train:]
+
+
+###############################################################################
+# Getting rules with a Decision Tree and Random Forests
+# ..................
+#
+# This part shows the training and performance evaluation of two tree-based models.
+# The objective remains to extract rules which targets credit defaults.
```
Comment (ngoix) [inline examples/plot_credit_default.py]: trailing whitespace

File: examples/plot_credit_default.py [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+fig, axes = plt.subplots(1, 2, figsize=(12, 5),
+                         sharex=True, sharey=True)
+matplotlib.rcParams.update({'font.size': 18})
+
+curves = [roc_curve, precision_recall_curve]
+xlabels = ['False Positive Rate', 'Recall (True Positive Rate)']
+ylabels = ['True Positive Rate (Recall)', 'Precision']
+
+for ax, curve, xlabel, ylabel in zip(axes.flatten(), curves,
+                                            xlabels, ylabels):
```
Comment (ngoix) [inline examples/plot_credit_default.py]: line over-indented

File: examples/plot_credit_default.py [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+    if curve == precision_recall_curve:
+        y_rf1, x_rf1, _ = curve(y_test, scoring_DT)
+        y_rf2, x_rf2, _ = curve(y_test, scoring_RF)
+        #y_rf3, x_rf3, _ = curve(y_test, scoring_ET)
+        ax.scatter(x_rf1, y_rf1, c='b', s=10,  label=label)
```
Comment (ngoix) [inline examples/plot_credit_default.py]: label not defined yet?

File: examples/plot_credit_default.py [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+# This part shows how FraudToRules can be fitted to detect credit defaults. Performances are compared with the random forest model previously trained.
+# 
+
+# fit the model
+rng = np.random.RandomState(42)
+
+clf = FraudToRules(similarity_thres = 1.0, max_depth=3, max_features=0.5, max_samples_features=0.5,
+                   random_state=rng, n_estimators=30,
+                   feature_names=feature_names, recall_min = 0.05, precision_min = 0.6
+                  )
```
Comment (ngoix) [inline examples/plot_credit_default.py]: closing bracket should be above for pep8 happiness

File: examples/plot_credit_default.py [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+print(str(len(clf.rules_)) + ' rules have been built.')
+print(clf.rules_[:5])
+
+###############################################################################
+# The most precise rules are displayed above.
+
+
+fig, axes = plt.subplots(1, 2, figsize=(12, 5),
+                         sharex=True, sharey=True)
+matplotlib.rcParams.update({'font.size': 18})
```
Comment (ngoix) [inline examples/plot_credit_default.py]: idem

File: examples/plot_credit_default.py [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+
+curves = [roc_curve, precision_recall_curve]
+xlabels = ['False Positive Rate', 'Recall (True Positive Rate)']
+ylabels = ['True Positive Rate (Recall)', 'Precision']
+
+for ax, curve, xlabel, ylabel in zip(axes.flatten(), curves,
+                                            xlabels, ylabels):
```
Comment (ngoix) [inline examples/plot_credit_default.py]: idem

File: examples/plot_credit_default.py [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+    if curve == precision_recall_curve:
+        y_rf1, x_rf1, _ = curve(y_test, scoring)
+        y_rf2, x_rf2, _ = curve(y_test, scoring_RF)
+        ax.scatter(x_rf1, y_rf1, c='b', s=10, label=label)
```
Comment (ngoix) [inline examples/plot_credit_default.py]: label not defined yet

File: examples/plot_credit_default.py [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+###############################################################################
+# Applying rules and predicting with ftrules
+# ..................
+#
+# This part shows how, once fitted, FraudToRules can be used to make predictions.
+# 
+
+
+###############################################################################
+# The end
```
Comment (ngoix) [inline examples/plot_credit_default.py]: rm thse 2 lines

**Comment by ngoix** [comment]
you'll have to rebase and push

File: examples/plot_credit_default.py [hunk 21]
```diff
@@ -0,0 +1,344 @@
...
+        url='https://archive.ics.uci.edu/ml/machine-learning-databases/'
+            '00350/default%20of%20credit%20card%20clients.xls',
+        checksum=('30c6be3abd8dcfd3e6096c828bad8c2f'
+                  '011238620f5369220bd60cfc82700933'))
+
+    if not exists(join(sk_data_dir, archive.filename)):
+        _fetch_remote(archive, dirname=sk_data_dir)
+
+    return(pd.read_excel(join(sk_data_dir, archive.filename),
+                         sheetname='Data', header=1))
```
Comment (ngoix) [inline examples/plot_credit_default.py]: do you have to return a pandas dataframe here? would be better to return a numpy array instead...

Comment (ngoix) [reply examples/plot_credit_default.py]: or a `Bunch` from sklearn/utils

File: examples/plot_credit_default.py [hunk 22]
```diff
@@ -0,0 +1,344 @@
...
+matplotlib.rcParams.update({'font.size': 18})
+
+curves = [roc_curve, precision_recall_curve]
+xlabels = ['False Positive Rate', 'Recall (True Positive Rate)']
+ylabels = ['True Positive Rate (Recall)', 'Precision']
+
+for ax, curve, xlabel, ylabel in zip(axes.flatten(),
+                                     curves, xlabels, ylabels):
+    if curve == precision_recall_curve:
+        y_rf1, x_rf1, _ = curve(y_test, scoring_DT)
```
Comment (ngoix) [inline examples/plot_credit_default.py]: x_dt x_rf instead of x_rf1 x_rf2?

File: examples/plot_credit_default.py [hunk 23]
```diff
@@ -0,0 +1,344 @@
...
+
+###############################################################################
+# Applying rules and predicting with skrules
+# ..................
+#
+# This part shows how, once fitted, SkopeRules can be used to
+# make predictions.
+
+###############################################################################
+# The end
```
Comment (ngoix) [inline examples/plot_credit_default.py]: rm thse 2 lines

File: examples/plot_credit_default.py [hunk 24]
```diff
@@ -0,0 +1,344 @@
...
+        _fetch_remote(archive, dirname=sk_data_dir)
+
+    return(pd.read_excel(join(sk_data_dir, archive.filename),
+                         sheetname='Data', header=1))
+
+# Importing data
+data = _load_credit_data()
+
+# Shuffling data, preparing target and variables
+data = data.sample(frac=1).reset_index(drop=True)
```
Comment (ngoix) [inline examples/plot_credit_default.py]: then you can use `shuffle` from sklearn.utils

File: examples/plot_credit_default.py [hunk 25]
```diff
@@ -0,0 +1,370 @@
...
+                         sharex=True, sharey=True)
+
+curves = [roc_curve, precision_recall_curve]
+xlabels = ['False Positive Rate', 'Recall (True Positive Rate)']
+ylabels = ['True Positive Rate (Recall)', 'Precision']
+
+for ax, curve, xlabel, ylabel in zip(axes.flatten(), curves,
+                                     xlabels, ylabels):
+    if curve == precision_recall_curve:
+        y_rf1, x_rf1, _ = curve(y_test, scoring)
```
Comment (ngoix) [inline examples/plot_credit_default.py]: idem here

File: examples/plot_credit_default.py [hunk 26]
```diff
@@ -0,0 +1,370 @@
...
+        y_rf2, x_rf2, _ = curve(y_test, scoring_RF)
+        ax.scatter(x_rf1, y_rf1, c='b', s=10, label=label)
+        ax.step(x_rf2, y_rf2, linestyle='-.', c='g', lw=1, where='post')
+        ax.set_title("Precision-Recall Curves", fontsize=20)
+    else:
+        x_rf1, y_rf1, _ = curve(y_test, scoring)
```
Comment (ngoix) [inline examples/plot_credit_default.py]: idem

File: examples/plot_credit_default.py [hunk 27]
```diff
@@ -0,0 +1,370 @@
...
+ylabels = ['True Positive Rate (Recall)', 'Precision']
+
+for ax, curve, xlabel, ylabel in zip(axes.flatten(), curves,
+                                     xlabels, ylabels):
+    if curve == precision_recall_curve:
+        y_rf1, x_rf1, _ = curve(y_test, scoring)
+        y_rf2, x_rf2, _ = curve(y_test, scoring_RF)
+        y_rf3, x_rf3, _ = curve(y_test, scoring_ET)
+        ax.scatter(x_rf1, y_rf1, c='b', s=10, label=label)
+        # ax.step(x_rf1, y_rf1, lw=2, where = 'post')
```
Comment (ngoix) [inline examples/plot_credit_default.py]: rm comments

File: examples/plot_skope_rules.py [hunk 28]
```diff
@@ -14,11 +14,13 @@
...
 
 """
-print(__doc__)
 
 import numpy as np
 import matplotlib.pyplot as plt
 from skrules import SkopeRules
+print(__doc__)
+
+print(__doc__)
```
Comment (ngoix) [inline examples/plot_skope_rules.py]: revert your changes here (due to rebase?)

File: skrules/datasets/credit_data.py [hunk 29]
```diff
@@ -0,0 +1,27 @@
+import pandas as pd
+import numpy as np
+from sklearn.datasets.base import get_data_home, Bunch
+from sklearn.datasets.base import _fetch_remote, RemoteFileMetadata
+from os.path import exists, join
+
+
+def _load_credit_data():
```
Comment (ngoix) [inline skrules/datasets/credit_data.py]: make it public

**Comment by ngoix** [comment]
LGTM +1 for merge ;)

**Comment by datajms** [comment]
LGTM too !
